fms_extras/modules/speculator.py

```diff
@@ -0,0 +1,91 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from fms.modules.layernorm import LayerNormParameterized
+
+class Speculator(nn.Module):
+    def __init__(self, emb_dim=4096, inner_dim=0, vocab_size=32000, n_predict=3):
+        super().__init__()
+        self.n_predict = n_predict
+        self.emb_dim = emb_dim
+        inner_dim = inner_dim if inner_dim!=0 else emb_dim
+        self.inner_dim = inner_dim
+        self.vsize = vocab_size
+        self.emb = nn.ModuleList([nn.Embedding(vocab_size, inner_dim) for _ in range(n_predict)])
+        self.proj = nn.ModuleList([nn.Linear((emb_dim if i==0 else inner_dim), inner_dim, bias=False) for i in range(n_predict)])
+        self.head = nn.ModuleList([nn.Linear(inner_dim, vocab_size, bias=False) for _ in range(n_predict)])
+        self.ln = nn.ModuleList(
+            [LayerNormParameterized(inner_dim, elementwise_shift=True, elementwise_scale=True) for _ in range(n_predict)]
+        )
+        # Weights ensure that state_0 accounts for 50% of state magnitude by final head in expectation
+        self.state_weight = .5**(.5/n_predict)
+        self.emb_weight = (1-self.state_weight**2)**.5
+        self.a = nn.GELU()
+
+    def reset_parameters(self):
+        for m in self.modules():
+            if isinstance(m, nn.Embedding) or isinstance(m, nn.Linear):
+                nn.init.trunc_normal_(m.weight, 0, 1 / self.inner_dim**0.5)
+            elif isinstance(m, LayerNormParameterized):
+                m.weight.data.fill_(1)
+                m.bias.data.zero_()
+
+    def generate_suffixes(self, state, ind, topk=[5, 4, 3], n=5):
```

Comment:
would multiple speculators share the same suffix-generation code? if so we might want to consider making this a separate utility function... though where ours takes "state" maybe it's really specific?

Reply:
Yeah, different speculators will have different suffix-generation functions because they'll likely take different inputs